google-cloud-spanner/lib/google/cloud/spanner/interval.rb

```diff
@@ -0,0 +1,270 @@
+
+          if days > MAX_DAYS || days < MIN_DAYS
+            raise ArgumentError, "The Interval class supports days from #{MIN_DAYS} to #{MAX_DAYS}."
+          end
+          @days = days
+
+          if nanoseconds > MAX_NANOSECONDS || nanoseconds < MIN_NANOSECONDS
+            raise ArgumentError, "The Interval class supports nanoseconds from #{MIN_NANOSECONDS} to #{MAX_NANOSECONDS}"
+          end
+          @nanoseconds = nanoseconds
+        end
+
+        def match_sign value
+          value.negative? ? -1 : 1
+        end
+
+        # Converts [Interval] to an ISO8601 Standard string.
+        def to_string
+          # Months should be converted to years and months.
+          years = @months.fdiv(12).truncate
+          months = @months % (match_sign(@months) * 12)
+
+          days = @days
+
+          # Nanoseconds should be converted to hours, minutes and seconds components.
+          remaining_nanoseconds = @nanoseconds
+
+          hours = (remaining_nanoseconds.abs / NANOSECONDS_IN_AN_HOUR) * match_sign(remaining_nanoseconds)
+          remaining_nanoseconds %= (match_sign(remaining_nanoseconds) * NANOSECONDS_IN_AN_HOUR)
+          minutes = (remaining_nanoseconds.abs / NANOSECONDS_IN_A_MINUTE) * match_sign(remaining_nanoseconds)
+          remaining_nanoseconds %= (match_sign(remaining_nanoseconds) * NANOSECONDS_IN_A_MINUTE)
+
+          # Only seconds can be fractional, and can have a maximum of 9 characters after decimal. Therefore,
+          # we convert the remaining nanoseconds to an integer for formatting.
+          seconds = (remaining_nanoseconds.abs / NANOSECONDS_IN_A_SECOND) * match_sign(remaining_nanoseconds)
+          nanoseconds = remaining_nanoseconds % (match_sign(remaining_nanoseconds) * NANOSECONDS_IN_A_SECOND)
+
+          interval_string = ["P"]
+
+          if years != 0
+            interval_string.append "#{years}Y"
+          end
+
+          if months != 0
+            interval_string.append "#{months}M"
+          end
+
+          if days != 0
+            interval_string.append "#{days}D"
+          end
+
+          if hours != 0 || minutes != 0 || seconds != 0 || nanoseconds != 0
+            interval_string.append "T"
+
+            if hours != 0
+              interval_string.append "#{hours}H"
+            end
+
+            if minutes != 0
+              interval_string.append "#{minutes}M"
+            end
+
+            if seconds != 0 || nanoseconds != 0
+              interval_string.append "#{format_seconds seconds, nanoseconds}S"
+            end
+          end
+
+          if interval_string == ["P"]
+            return "P0Y"
+          end
+
+          interval_string.join
+        end
+
+        # Formats decimal values be in multiples of 3 length.
+        #
+        def format_seconds seconds, nanoseconds
+          return seconds if nanoseconds.zero?
+          add_sign = seconds.zero? && nanoseconds.negative?
+
+          nanoseconds_str = nanoseconds.abs.to_s.rjust 9, "0"
```

Comment:
Why do you need the rjust if you're just going to remove the trailing zeros in the next line?

Reply:
oh, right, of course. /me getting left and right confused again...